tests/test_triples_factory.py

```diff
@@ -251,45 +255,45 @@ def test_split_multi(self):
 
     def test_cleanup_deterministic(self):
         """Test that triples in a test set can get moved properly to the training set."""
-        training = np.array([
+        training = torch.as_tensor(data=[
             [1, 1000, 2],
             [1, 1000, 3],
             [1, 1001, 3],
-        ])
-        testing = np.array([
+        ], dtype=torch.long)
+        testing = torch.as_tensor(data=[
             [2, 1001, 3],
             [1, 1002, 4],
-        ])
-        expected_training = [
+        ], dtype=torch.long)
+        expected_training = torch.as_tensor(data=[
             [1, 1000, 2],
             [1, 1000, 3],
             [1, 1001, 3],
             [1, 1002, 4],
-        ]
-        expected_testing = [
+        ], dtype=torch.long)
+        expected_testing = torch.as_tensor(data=[
             [2, 1001, 3],
-        ]
+        ], dtype=torch.long)
 
         new_training, new_testing = _tf_cleanup_deterministic(training, testing)
-        self.assertEqual(expected_training, new_training.tolist())
-        self.assertEqual(expected_testing, new_testing.tolist())
+        assert (expected_training == new_training).all()
+        assert (expected_testing == new_testing).all()
 
         new_testing, new_testing = _tf_cleanup_all([training, testing])
-        self.assertEqual(expected_training, new_training.tolist())
-        self.assertEqual(expected_testing, new_testing.tolist())
+        assert (expected_training == new_training).all()
```

Comment:
you can use `self.assertTrue`. The only reason I've switched a lot of assert statements into these functional forms is that unittest/pytest will give more information/context on failures

Reply:
For me it is exactly the opposite:
With self.assert*, I get a fancy comparison between expected and the actual value, but I cannot directly see the stack trace of the failed comparison. With "assert ..." I get a full stack trace of all participating methods. Maybe that is due to having `pytest` as default test runner?

Reply:
oh boy... Are you checking the tests using tox or an IDE? Tox's configuration uses pytest, but I think typically I use unittest in the IDE becuase it integrates better.

Reply:
IDE, Pycharm in particular. It actually might come from PyCharm magic things.

That's how they look like

```ptb
FAILED [ 14%]
 
{'a': 0, 'a.': 1, 'a._inverse': 2} != {'a': 0, 'a.': 2, 'a._inverse': 3, 'a_inverse': 1}

Expected :{'a': 0, 'a.': 2, 'a._inverse': 3, 'a_inverse': 1}
Actual   :{'a': 0, 'a.': 1, 'a._inverse': 2}
<Click to see difference>

self = <tests.test_triples_factory.TestTriplesFactory testMethod=test_automatic_incomplete_inverse_detection>
first = {'a': 0, 'a.': 2, 'a._inverse': 3, 'a_inverse': 1}
second = {'a': 0, 'a.': 1, 'a._inverse': 2}, msg = None

    def _patched_equals(self, first, second, msg=None):
        try:
>           old(self, first, second, msg)

/snap/pycharm-professional/226/plugins/python/helpers/pycharm/teamcity/diff_tools.py:32: 
```

Reply:
I agree with @mberr. I use the pytest integration in Pycharm as well and `assert` gives a much more sensible stack trace of the failed comparison. It actually shows exactly where the error lies in the code no matter how many inheritances or function calls one is away from the test call itself. `self.assertTrue` only mentions the failed test, but only shows a stack trace to where the comparison function itself failed (not where the result actually comes from), which is interesting but not informative at all 🤷‍♂️

Reply:
@cthoyt Do we actually have any documentation on our "testing strategy" and with which design considerations our tests were setup?

Reply:
No, there isn't a place documenting the testing style. @lvermue I'm surprised you're having the same problem as Max ☹️